codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedStringGenerator.kt

```diff
@@ -145,39 +163,172 @@ class ConstrainedStringGenerator(
-                    rustBlockTemplate(
-                        "pub(crate) fn as_validation_exception_field(self, path: #{String}) -> crate::model::ValidationExceptionField",
-                        "String" to RuntimeType.String,
-                    ) {
-                        rustBlock("match self") {
-                            rust(
-                                """
-                                Self::Length(length) => crate::model::ValidationExceptionField {
-                                    message: format!("${lengthTrait.validationErrorMessage()}", length, &path),
-                                    path,
-                                },
-                                """,
-                            )
-                        }
+private data class Pattern(val patternTrait: PatternTrait) : StringTraitInfo() {
+    override fun toTraitInfo(): TraitInfo {
+        val pattern = patternTrait.pattern
+
+        return TraitInfo(
+            { rust("let value = Self::check_pattern(value)?;") },
+            {
+                docs("Error when a string doesn't satisfy its `@pattern`.")
+                docs("Contains the String that failed the pattern.")
+                rust("Pattern(String),")
+            },
+            {
+                rust(
+                    """
+                    Self::Pattern(string) => crate::model::ValidationExceptionField {
+                        message: format!("${patternTrait.validationErrorMessage()}", &string, &path, r##"$pattern"##),
+                        path
+                    },
+                    """,
+                )
+            },
+            this::renderValidationFunction,
+        )
+    }
+
+    /**
+     * Renders a `check_pattern` function to validate the string matches the
+     * supplied regex in the `@pattern` trait.
+     */
+    private fun renderValidationFunction(constraintViolation: Symbol): Writable {
+        val pattern = patternTrait.pattern
+
+        return {
+            rustTemplate(
+                """
+                fn check_pattern(string: String) -> Result<String, $constraintViolation> {
+                    let regex = Self::compile_regex();
+
+                    if regex.is_match(&string) {
+                        Ok(string)
+                    } else {
+                        Err($constraintViolation::Pattern(string))
                     }
                 }
-            }
+
+                pub fn compile_regex() -> &'static #{Regex}::Regex {
+                    static REGEX: #{OnceCell}::sync::OnceCell<#{Regex}::Regex> = #{OnceCell}::sync::OnceCell::new();
+
+                    REGEX.get_or_init(||
+                        #{Regex}::Regex::new(r##"$pattern"##)
+                            .unwrap_or_else(|_| panic!("The regular expression {} is not supported by the `regex` crate; feel free to file an issue under https://github.com/awslabs/smithy-rs/issues for support", r##"$pattern"##))
```

Comment:
Why not just `unwrap_or`? The message is `&'static`.

Reply:
I think you mean `expect`, but you're right. Fixed in https://github.com/awslabs/smithy-rs/pull/1998/commits/fdf88b0de0cae3a4c33c6e4105dbddfb0a9b25dc.